File: minterpreters/5-part10-cpp/src/parser.cpp
#include <memory>

#include "ast_visitor.h"
#include "lexer.h"
#include "parser.h"
#include "token.h"

Parser::Parser(std::string user_code) : lexer(std::make_unique<Lexer>(user_code)) {}

Parser::~Parser() {};

std::shared_ptr<AST> Parser::parse() {
    Token token = lexer->get_current_token();
    std::shared_ptr<AST> node = program();

    return node;
}

void Parser::parsing_error(std::string err_message) {
    if (err_message == "") {
        std::cout << "Parser error" << std::endl;
    }
    else {
        std::cout << err_message << std::endl;
        exit(EXIT_FAILURE);
    }
}

void Parser::eat(std::string token_type) {
    std::string current_token_type = lexer->get_current_token().get_type();
    if (current_token_type == token_type) {
        lexer->get_next_token();
    }
    else {
        parsing_error("Failed eat request, got: "+ current_token_type + 
                      " expected: " + token_type);
    }
}

/*
* GRAMMARS
*/

std::shared_ptr<AST> Parser::program() {
    /*
     * PROGRAM VARIABLE SEMI BLOCK .
    */
    std::shared_ptr<AST> node;
    eat("PROGRAM");
    eat("VARIABLE");
    eat("SEMI");
    node = block();
    eat("DOT");

    return node;
}

std::shared_ptr<AST> Parser::block() {
    /*
     * DECLARATIONS COMPOUND_STATEMENT
    */
    std::shared_ptr<AST> block_node;
    std::shared_ptr<AST> declaration_node;
    std::shared_ptr<AST> cs_node;

    if (lexer->get_current_token().get_type() == "VAR") {
        declaration_node = declarations();
    }

    cs_node = compound_statement();

    block_node = std::make_shared<BlockNode> (BlockNode(declaration_node, cs_node));
    return block_node;
}

std::shared_ptr<AST> Parser::declarations() {
    /*
     * VAR (variable_declarations SEMI)+ | empty
    */
    std::shared_ptr<AST> node;
    eat("VAR");

    if (lexer->get_current_token().get_type() != "VARIABLE") {
        node = empty(); 
    }

    while (lexer->get_current_token().get_type() == "VARIABLE") {
        variable_declarations();
        eat("SEMI");
    }

    return node;
}

std::shared_ptr<AST> Parser::variable_declarations() {
    /*
     * ID (COMMA ID)* COLON type_spec
    */
    std::shared_ptr<AST> node;
    eat("VARIABLE");

    while (lexer->get_current_token().get_type() == "COMMA") {
        eat("COMMA");
        eat("VARIABLE");
    }

    eat("COLON");
    type_spec();

    return node;
}

std::shared_ptr<AST> Parser::type_spec() {
    /*
     * INTEGER | REAL
    */    
    std::shared_ptr<AST> node;
    eat("TYPE");

    return node;
}


std::shared_ptr<AST> Parser::compound_statement() {
    /*
     * BEGIN STATEMENT_LIST END
    */
    eat("BEGIN");

    std::shared_ptr<AST> node;
    node = statement_list();

    eat("END");
    return node;

}

std::shared_ptr<AST> Parser::statement_list() {
    /*
     * STATEMENT*
    */
    std::unique_ptr<CompoundNode> node = std::make_unique<CompoundNode>();
    
    node->add_statement( statement() );
    while (lexer->get_current_token().get_type() != "END") {
        node->add_statement( statement() );
    }

    return node;
}

std::shared_ptr<AST> Parser::statement() {
    /*
     * (COMPOUND STATEMENT | ASSIGN_OP | EMPTY) SEMI
    */
    std::shared_ptr<AST> node;
    std::string node_type = lexer->get_current_token().get_type();

    if (node_type == "TYPE") {node = assign_op();}
    else if (node_type == "BEGIN") {node = compound_statement();}
    else {node = empty();}

    eat("SEMI");

    return node;
}

std::shared_ptr<AST> Parser::assign_op() {
    /*
    * TYPE VARIABLE EQUALS EXPR
    */
    // TYPE
    Token type_token = lexer->get_current_token();
    std::shared_ptr<AST> type_node = std::make_shared<TypeNode>(type_token.get_type(), type_token.get_value());
    eat("TYPE");

    // VARIABLE
    Token var_token = lexer->get_current_token();
    std::shared_ptr<AST> var_node = std::make_shared<VariableNode>(var_token);
    eat("VARIABLE");
    eat("EQUALS");

    // EXPR
    std::shared_ptr<AST> value_node (expr());

    std::shared_ptr<AST> assign_node = std::make_shared<AssignNode>(type_node, var_node, value_node);
    return assign_node;
}

std::shared_ptr<AST> Parser::expr() {
    /*
    * TERM ((PLUS|MINUS) TERM)*
    */   
    std::shared_ptr<AST> node (term());

    Token op = lexer->get_current_token();
    while((op.get_type() == "PLUS") || op.get_type() == "MINUS" ) {
        if (op.get_type()=="PLUS") { eat("PLUS"); }
        else { eat("MINUS"); }

        node = std::make_shared<BinOpNode>(BinOpNode(node, op, term()));
        op = lexer->get_current_token();
    }

    return node;
}

std::shared_ptr<AST> Parser::term() {
    /*
     * FACTOR ((MUL|DIV) FACTOR)*
     */
    std::shared_ptr<AST> node (factor());

    Token op = lexer->get_current_token();
    while((op.get_type() == "MUL") || op.get_type() == "DIV" ) {
        if (op.get_type()=="MUL") {
            eat("MUL");
        }
        else {
            eat("DIV");
        }

        node = std::make_shared<BinOpNode>(BinOpNode(node, op, factor()));
        op = lexer->get_current_token();
    }

    return node;
}

std::shared_ptr<AST> Parser::factor() {
    /*
    * (UNARY + | UNARY -)factor | INTEGER | REAL | LPAREN EXPR RPAREN
    */
    Token token = lexer->get_current_token();
    std::shared_ptr<AST> node;

    if (token.get_type() == "PLUS") {
        eat("PLUS");
        node = factor();
        node = std::make_shared<UnaryNode>(UnaryNode("PLUS", node));
    } 
    else if (token.get_type() == "MINUS") {
        eat("MINUS");
        node = factor();
        node = std::make_shared<UnaryNode>(UnaryNode("MINUS", node));
    } 
    else if (token.get_type() == "INTEGER") {
        int token_value = std::stoi(token.get_value());
        node = std::make_shared<IntegerNode>(IntegerNode(token_value));
        eat("INTEGER");
    } 
    else if (token.get_type() == "REAL") {
        float token_value = std::stof(token.get_value());
        node = std::make_shared<RealNode>(RealNode(token_value));
        eat("REAL");
    }
    else if (token.get_type() == "LPAREN") {
        eat("LPAREN");
        node = expr();
        eat("RPAREN");
    }
    else {
        parsing_error("No matching options found in 'factor' grammar");
    }

    return node;   
}

std::shared_ptr<AST> Parser::empty() {
    std::shared_ptr<EmptyNode> node = std::make_shared<EmptyNode>(EmptyNode());

    return node;
}

File: minterpreters/part15/parser.py
from errors import Error, ErrorCode, ParserError
from parser_objects import (AssignNode, BinOpNode, BlockNode, CompoundNode, IntegerConstNode, 
                            IntegerNode, NoOp, ParamNode, ProcedureDeclNode, ProgramNode, 
                            RealConstNode, RealNode, UnaryNode, VarDeclNode, VarNode)
from tokens import Token, TokenType, RESERVED_KEYWORDS

_SHOULD_LOG_SCOPE = True

###############################################################################
#                                                                             #
#  PARSER                                                                     #
#                                                                             #
###############################################################################

'''
GRAMMERS:

program: PROGRAM variable SEMI block DOT
block: declarations compound_statement
declarations :  
    (VAR (declaration SEMI)+)*
    | (PROCEDURE ID (LPAREN formal_parameter_list RPAREN)? SEMI block SEMI)*
    | empty
declaration: variable (COMMA variable) COLON type_spec SEMI
formal_parameter_list: 
    formal_parameters
    | formal_parameters SEMI formal_parameter_list
formal_parameters : ID (COMMA ID)* COLON type_spec
compound_statement: BEGIN statement_list END
statement_list: statement | statement SEMI statement_list
statement: compound_statement | assign_statement | empty
empty: 
assign_statement: variable EQUALS expr
expr: term ((PLUS | MINUS) term)*
term: factor ((MUL | INTEGER_DIV | FLOAT_DIV) factor)*  
factor: INTEGER_CONST | REAL_CONST | LPAREN expr RPAREN
variable: ID
type_spec: INTEGER | REAL
'''

class Parser:
    def __init__(self, lexer):
        self.lexer = lexer
        self.current_token = lexer.get_next_token()
        self.log(f'Next token: {self.current_token}')

    def eat(self, token_type):
        if self.current_token.type == token_type:
            self.current_token = self.lexer.get_next_token()
            self.log(f'Next token: {self.current_token}')
        else:
            self.error(
                error_code = ErrorCode.UNEXPECTED_TOKEN,
                token = self.current_token,
                expected = token_type,
            )

    def error(self, error_code, token, expected=None):
        message = f'{error_code.value} -> {token}'
        if expected: message += f' Expected -> {expected}'
        raise ParserError(
            error_code = error_code,
            token = token,
            message = message
        )

    def log(self, msg):
        if _SHOULD_LOG_SCOPE:
            print(msg)

    def parse(self):
        return self.program()           

    def program(self):
        #program: PROGRAM variable SEMI block DOT
        self.eat(TokenType.PROGRAM)
        var_name = self.current_token.value
        self.eat(TokenType.ID)
        self.eat(TokenType.SEMI)
        node = self.block()
        self.eat(TokenType.DOT)
        self.log("Parsing complete")
        return ProgramNode(var_name, node)

    def block(self):
        #block: declarations compound_statement
        decl_node = self.declarations()       
        cs_node = self.compound_statement()
        return BlockNode(decl_node, cs_node)    

    def declarations(self):
        '''
        declarations :  
            (VAR (variable_declaration SEMI)+)* procedure_declaration*
            | empty
        '''
        declarations = []
        if self.current_token.type == TokenType.VAR:
            self.eat(TokenType.VAR)
            while self.current_token.type == TokenType.ID:
                declarations.extend(self.declaration())
                self.eat(TokenType.SEMI)
        
        while self.current_token.type == TokenType.PROCEDURE:
            proc_decl = self.procedure_declaration()
            declarations.append(proc_decl)
        
        return declarations       

    def declaration(self):
        #declaration: variable (COMMA variable) COLON type
        variables = [self.variable()]
        while self.current_token.type == TokenType.COMMA:
            self.eat(TokenType.COMMA)
            variables.append(self.variable())
    
        self.eat(TokenType.COLON)
        type_node = self.type_spec()

        var_declarations = []
        for var_node in variables:
            var_declarations.append(VarDeclNode(var_node, type_node))

        return var_declarations

    def procedure_declaration(self):
        '''
        PROCEDURE ID (LPAREN formal_parameter_list RPAREN)? SEMI block SEMI
        '''
        self.eat(TokenType.PROCEDURE)
        name = self.current_token.value
        self.eat(TokenType.ID)
        parameter_list = None

        if self.current_token.type == TokenType.LPAREN:
            self.eat(TokenType.LPAREN)
            parameter_list = self.formal_parameter_list()
            self.eat(TokenType.RPAREN)
        
        self.eat(TokenType.SEMI)
        block = self.block()
        self.eat(TokenType.SEMI)    

        return ProcedureDeclNode(name, parameter_list, block)

    def formal_parameter_list(self):
        '''
        formal_parameter_list: 
            formal_parameters
            | formal_parameters SEMI formal_parameter_list
        '''
        formal_parameters = self.formal_parameters()
        if self.current_token.type == TokenType.SEMI:
            self.eat(TokenType.SEMI)
            formal_parameters.extend(self.formal_parameter_list())

        return formal_parameters

    def formal_parameters(self):
        '''
        formal_parameters : ID (COMMA ID)* COLON type_spec
        '''
        variables = [VarNode(self.current_token)]
        self.eat(TokenType.ID)
        while self.current_token.type == TokenType.COMMA:
            self.eat(TokenType.COMMA)
            variables.append(VarNode(self.current_token))
            self.eat(TokenType.ID)

        self.eat(TokenType.COLON)
        type_node = self.type_spec()

        param_list = []
        for var_node in variables:
            param_list.append(ParamNode(var_node, type_node))

        return param_list

    def type_spec(self):
        #type_spec: INTEGER | REAL
        token = self.current_token
        if token.type == TokenType.INTEGER:
            self.eat(TokenType.INTEGER)
            return IntegerNode(token)
        elif token.type == TokenType.REAL:
            self.eat(TokenType.REAL)
            return RealNode(token)

    def compound_statement(self):
        #compound_statement: BEGIN statement_list END
        self.eat(TokenType.BEGIN)
        statements = self.statement_list()
        self.eat(TokenType.END)
        return CompoundNode(statements)    

    def statement_list(self):
        #statement_list: statement | statement SEMI statement_list
        statement_list = [self.statement()]
        while self.current_token.type == TokenType.SEMI:
            self.eat(TokenType.SEMI)
            statement_list.append(self.statement())
        return statement_list      

    def statement(self):
        #statement: compound_statement | assign_statement | empty
        if self.current_token.type == TokenType.BEGIN:
            node = self.compound_statement()
        elif self.current_token.type == TokenType.ID:
            node = self.assign_statement()
        else:
            node = self.empty()
        return node       

    def empty(self):
        #empty: 
        return NoOp()    

    def assign_statement(self):
        #assign_statement: variable EQUALS expr
        var_name = self.variable()
        self.eat(TokenType.ASSIGN)
        value = self.expr()

        node = AssignNode(var_name, value)
        return node             

    def expr(self):
        #expr: term ((PLUS | MINUS) term)*
        node = self.term()
        while self.current_token.type in [TokenType.PLUS, TokenType.MINUS]:
            op = self.current_token
            if op.type == TokenType.PLUS:
                self.eat(TokenType.PLUS)
            elif op.type == TokenType.MINUS:
                self.eat(TokenType.MINUS)
            right = self.term()
            node = BinOpNode(node, op, right)        
        return node

    def term(self):
        #term: factor ((MUL | INTEGER_DIV | FLOAT_DIV) factor)*         
        node = self.factor()
        while self.current_token.type in [TokenType.MUL, TokenType.INTEGER_DIV, TokenType.FLOAT_DIV]:
            op = self.current_token
            if op.type == TokenType.MUL:
                self.eat(TokenType.MUL)
            elif op.type == TokenType.INTEGER_DIV:
                self.eat(TokenType.INTEGER_DIV)
            elif op.type == TokenType.FLOAT_DIV:
                self.eat(TokenType.FLOAT_DIV)
            right = self.factor()
            node = BinOpNode(node, op, right)
        return node

    def factor(self):
        #factor: unary (INTEGER_CONST | REAL_CONST | variable) | LPAREN expr RPAREN
        token = self.current_token
        if token.type == TokenType.PLUS:
            self.eat(TokenType.PLUS)
            node = UnaryNode(TokenType.PLUS, self.factor())
        elif token.type == TokenType.MINUS:
            self.eat(TokenType.MINUS)   
            node = UnaryNode(TokenType.MINUS, self.factor())                 
        elif token.type == TokenType.INTEGER_CONST:
            self.eat(TokenType.INTEGER_CONST)
            node = IntegerConstNode(token)
        elif token.type == TokenType.REAL_CONST:
            self.eat(TokenType.REAL_CONST)
            node = RealConstNode(token)
        elif token.type == TokenType.LPAREN:
            self.eat(TokenType.LPAREN)
            node = self.expr()
            self.eat(TokenType.RPAREN)
        elif token.type == TokenType.ID:
            node = self.variable()
        else:
            self.error(
                error_code = ErrorCode.UNEXPECTED_TOKEN,
                token = token
            )

        return node

    def variable(self):
        #variable: ID
        node = VarNode(self.current_token)
        self.eat(TokenType.ID)
        return node
